### PR TITLE
feat(block-producer): query contracts cell dep on chain

### PR DESCRIPTION
**Pull request by zeroqn**

### CODE REVIEW

File: crates/rpc-client/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        .build();
+    let type_script_hash = ckb_fixed_hash::H256(type_script_hash.into()).pack();
+
+    let search_key = SearchKey {
+        script: contract_type_script.into(),
+        script_type: ScriptType::Type,
+        filter: None,
+    };
+    let order = Order::Desc;
+    let limit = Uint32::from(DEFAULT_QUERY_LIMIT as u32);
```
Comment (jjyr) [inline crates/rpc-client/src/contract.rs]: we only need to query 1 cell

**Review by jjyr — CHANGES_REQUESTED** [review]

**Comment by jjyr** [comment]
We should perform the searching before produce a block, so if the contract is updated, we could automatically update the reference.

**Comment by zeroqn** [comment]
> We should perform the searching before produce a block, so if the contract is updated, we could automatically update the reference.

Done

File: crates/block-producer/src/runner.rs [hunk 2]
```diff
@@ -232,9 +236,12 @@ pub struct BaseInitComponents {
     pub rpc_client: RPCClient,
     pub store: Store,
     pub generator: Arc<Generator>,
+    pub opt_block_producer_config: Option<BlockProducerConfig>,
+    pub contracts_dep_manager: Option<ContractsCellDepManager>,
 }
 
 impl BaseInitComponents {
+    #[allow(deprecated)]
```
Comment (jjyr) [inline crates/block-producer/src/runner.rs:243]: which function is deprecated?

File: crates/block-producer/src/runner.rs [hunk 3]
```diff
@@ -433,17 +480,19 @@ pub fn run(config: Config, skip_config_check: bool) -> Result<()> {
     );
 
     let base = BaseInitComponents::init(&config, skip_config_check)?;
+    config.block_producer = base.opt_block_producer_config.clone();
```
Comment (jjyr) [inline crates/block-producer/src/runner.rs]: Why?, the config should be a immutable structure

File: crates/config/src/config.rs [hunk 4]
```diff
@@ -109,26 +109,81 @@ pub struct ChallengerConfig {
...
+    #[serde(default)]
     pub allowed_eoa_deps: HashMap<H256, CellDep>,
+    #[deprecated]
+    #[serde(skip_serializing)]
+    #[serde(default)]
     pub allowed_contract_deps: HashMap<H256, CellDep>,
     pub challenger_config: ChallengerConfig,
     pub wallet_config: WalletConfig,
+    #[serde(default)]
+    pub contract_type_scripts: ContractTypeScriptConfig,
```
Comment (jjyr) [inline crates/config/src/config.rs]: This part is belongs to consensus, we I think maybe we can move it out from BlockProducer, and add a new consensus config?

``` rust
pub struct ConsensusConfig {
  pub contract_type_scripts: ContractTypeScriptConfig,
}
```

File: crates/block-producer/src/runner.rs (hunk 2 above)
Comment (zeroqn) [reply crates/block-producer/src/runner.rs:243]: ```fn query_type_script_from_old_config()```

**Review by jjyr — DISMISSED** [review]

File: crates/rpc-client/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+impl ContractsCellDepManager {
+    pub async fn build(rpc_client: RPCClient, scripts: ContractTypeScriptConfig) -> Result<Self> {
+        let now = Instant::now();
+        let deps = query_cell_deps(&rpc_client, &scripts).await?;
+        log::trace!("[contracts dep] build {}ms", now.elapsed().as_millis());
+
+        Ok(Self {
+            rpc_client,
+            scripts: Arc::new(scripts),
+            deps: Arc::new(ArcSwap::new(Arc::new(deps))),
```
Comment (magicalne) [inline crates/rpc-client/src/contract.rs]: This line makes me think for 20min at least... I suspect both "Arc"s are not necessary. 
I guess we can just use a normal pointer here guarded by ArcSwap.
```rust
deps: ArcSwap::from_pointee(deps);
```

Comment (zeroqn) [reply crates/rpc-client/src/contract.rs]: Done

File: crates/rpc-client/src/contract.rs [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+use crate::indexer_types::{Cell, Order, Pagination, ScriptType, SearchKey};
+use crate::rpc_client::RPCClient;
+use crate::utils::to_result;
+
+// Used in block producer and challenge
+#[derive(Clone)]
+pub struct ContractsCellDepManager {
+    rpc_client: RPCClient,
+    scripts: Arc<ContractTypeScriptConfig>,
+    deps: Arc<ArcSwap<ContractsCellDep>>,
```
Comment (magicalne) [inline crates/rpc-client/src/contract.rs:23]: This Arc can also get removed. Elaborately **load()** and **store()** get auto deref twice here. I guess ArcSwap guards the pointee with AtoimicPointer with Strategy internally. So Arc takes no effect here?

```suggestion
    deps:ArcSwap<ContractsCellDep>,
```
This should be fine.

File: crates/rpc-client/src/contract.rs [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+impl ContractsCellDepManager {
+    pub async fn build(rpc_client: RPCClient, scripts: ContractTypeScriptConfig) -> Result<Self> {
+        let now = Instant::now();
+        let deps = query_cell_deps(&rpc_client, &scripts).await?;
+        log::trace!("[contracts dep] build {}ms", now.elapsed().as_millis());
+
+        Ok(Self {
+            rpc_client,
+            scripts: Arc::new(scripts),
+            deps: Arc::new(ArcSwap::from_pointee(deps)),
```
Comment (magicalne) [inline crates/rpc-client/src/contract.rs:35]:
```suggestion
            deps: ArcSwap::from_pointee(deps),
```

File: crates/rpc-client/src/contract.rs [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    }
+
+    pub async fn refresh(&self) -> Result<()> {
+        log::info!("[contracts dep] refresh");
+
+        let now = Instant::now();
+        let deps = query_cell_deps(&self.rpc_client, &self.scripts).await?;
+        log::trace!("[contracts dep] refresh {}ms", now.elapsed().as_millis());
+
+        self.deps.store(Arc::new(deps));
```
Comment (magicalne) [inline crates/rpc-client/src/contract.rs:50]:
```suggestion
self.deps.store(deps);
```

File: crates/rpc-client/src/contract.rs (hunk 6 above)
Comment (zeroqn) [reply crates/rpc-client/src/contract.rs:23]: No. ArcSwap doesn't implement Clone.

File: crates/rpc-client/src/contract.rs (hunk 8 above)
Comment (zeroqn) [reply crates/rpc-client/src/contract.rs:50]: Compile failure. Different type.

Comment (magicalne) [reply crates/rpc-client/src/contract.rs:50]: You're right. This is fine.

**Review by magicalne — DISMISSED** [review]

**Review by jjyr — DISMISSED** [review]
Pls rebase

**Review by jjyr — APPROVED** [review]

**Pull request merged**